File: 1001~1500/1143. Longest Common Subsequence/main.cpp
// Source: https://leetcode.com/problems/longest-common-subsequence/
// Date: 2021/7/10
// Dynamic Programming LCS Problem

#include<iostream>
#include<string>
using namespace std;

class Solution {
public:
    int ans[1001][1001] = {0};
    int longestCommonSubsequence(string text1, string text2) {
        int l1 = text1.length(), l2 = text2.length();
        for(int i=1; i<=l1; i++) {
            for(int j=1; j<=l2; j++) {
                if(text1[i-1] == text2[j-1])
                    ans[i][j] = ans[i-1][j-1]+1;
                else
                    ans[i][j] = max(ans[i-1][j], ans[i][j-1]);
            }
        }
        return ans[l1][l2];
    }


};

int main() {
    string s1 = "bsbininm";
    string s2 = "jmjkbkjkv";
    Solution s;
    s.longestCommonSubsequence(s1,s2);
    for(int i=1; i<=8; i++) {
        for(int j=1; j<=9; j++) 
            cout<<s.ans[i][j]<<' ';
        cout<<endl;
    }    
}



static const auto io_sync_off = []() {
    // turn off sync
    std::ios::sync_with_stdio(false);
    // untie in/out streams
    std::cin.tie(nullptr);
    return nullptr;
}();

File: 1001~1500/1143. Longest Common Subsequence/main.py
"""
Source: https://leetcode.com/problems/longest-common-subsequence/description/
Date: 2022/12/15
Skill:
Runtime: 423 ms, faster than 90.30% 
Memory Usage: 22 MB, less than 81.56%
Time complexity: 
Space complexity: 
Constraints: 
    1 <= text1.length, text2.length <= 1000
    text1 and text2 consist of only lowercase English characters.
"""

from typing import List, Optional

class Solution:
    def longestCommonSubsequence(self, text1: str, text2: str) -> int:
        dp = [[0]*(len(text1)+1) for i in range(len(text2)+1)]
        for i in range (1, len(text2)+1):
            for j in range (1, len(text1)+1):
                if text2[i-1] == text1[j-1]:
                    dp[i][j] = dp[i-1][j-1]+1
                else:
                    dp[i][j] = max(dp[i-1][j], dp[i][j-1])
        
        return dp[len(text2)+1][len(text1)+1]